packages/graphql-eslint-rules/src/missing-apollo-key-fields.ts

```diff
@@ -0,0 +1,263 @@
+    );
+  }
+}
+
+function keyFieldsForType(
+  type: GraphQLObjectType | GraphQLInterfaceType,
+  typePolicies: any
+) {
+  const keyFields: string[] = [];
+  const typePolicy = typePolicies[type.name];
+  if (typePolicy && typePolicy.keyFields) {
+    if (Array.isArray(typePolicy.keyFields)) {
+      typePolicy.keyFields.forEach((keyField: any) => {
+        if (typeof keyField === "string") {
+          keyFields.push(keyField);
+        } else {
+          throw new Error("Expected keyFields to be array of strings");
+        }
+      });
+    } else {
+      throw new Error("Expected keyFields to be array of strings");
+    }
+  } else if (type.getFields().id !== undefined) {
+    keyFields.push(DEFAULT_KEY_FIELD_NAME);
+  }
+  return keyFields;
+}
+
+function getKeyFieldsObjectForCheck(keyFields: string[]) {
+  return keyFields.reduce((acc, id) => {
+    acc[id] = false;
+    return acc;
+  }, {} as Record<string, boolean>);
+}
+
+function getUnusedKeyFields(idsObjectForCheck: Record<string, boolean>) {
+  return Object.entries(idsObjectForCheck).reduce((acc, [key, value]) => {
+    if (value) {
+      return acc;
+    }
+    acc.push(key);
+    return acc;
+  }, [] as string[]);
+}
+
+export default {
+  meta: {
+    type: "suggestion",
+    docs: {
+      category: "Best Practices",
+      description: `Enforce selecting specific fields when they are available on the GraphQL type.`,
+      requiresSchema: true,
+      requiresSiblings: true,
+      examples: [
+        {
+          title: "Incorrect",
+          code: /* GraphQL */ `
+            # In your schema
+            type User {
+              id: ID!
+              name: String!
+            }
+            # Query
+            query user {
+              user {
+                name
+              }
+            }
+          `,
+        },
+        {
+          title: "Correct",
+          code: /* GraphQL */ `
+            # In your schema
+            type User {
+              id: ID!
+              name: String!
+            }
+            # Query
+            query user {
+              user {
+                id
+                name
+              }
+            }
+          `,
+        },
+      ],
+    },
+    messages: {
+      [REQUIRE_KEY_FIELDS_WHEN_AVAILABLE]: `Field "{{ fieldName }}" must be selected when it's available on a type. Please make sure to include it in your selection set!\nIf you are using fragments, make sure that all used fragments {{checkedFragments}} specifies the field "{{ fieldName }}".`,
+    },
+    schema: {
+      type: "array",
+      additionalItems: false,
+      minItems: 0,
+      maxItems: 1,
+      items: {
+        type: "object",
+        properties: {
+          fieldName: {
+            type: "string",
+            default: DEFAULT_KEY_FIELD_NAME,
+          },
+        },
+      },
+    },
+  },
+  create(context: any) {
+    return {
+      SelectionSet(node: any) {
```

Comment:
```suggestion
      SelectionSet(node: SelectionSetNode) {
```

Reply:
Doesn't work. I think the graphql-eslint-plugin enhance the original graphql `SelectionSetNode`